cmd/gcp-controller-manager/dpwi/serviceaccounts/verifier.go

```diff
@@ -0,0 +1,198 @@
+package serviceaccounts
+
+import (
+	"fmt"
+	"sync"
+
+	"golang.org/x/sync/singleflight"
+	core "k8s.io/api/core/v1"
+	coreinformers "k8s.io/client-go/informers/core/v1"
+	"k8s.io/client-go/tools/cache"
+	clientcmdapi "k8s.io/client-go/tools/clientcmd/api"
+	"k8s.io/cloud-provider-gcp/cmd/gcp-controller-manager/dpwi/hms"
+	"k8s.io/klog/v2"
+)
+
+const (
+	// ServiceAccountAnnotationGSAEmail is the key to GCP Service Account annotation in
+	// ServiceAccount objects.
+	ServiceAccountAnnotationGSAEmail = "iam.gke.io/gcp-service-account"
+)
+
+// ServiceAccountVerifier implements a custom control loop responsible for verifying authorization
+// for Kubernetes Service Accounts (KSA) in the cluster to impersonate specific GCP Service Accounts
+// (GSA) and persisting the authorized pairs in a ConfigMap entry.
+//
+// verifiedSAs is a thread-safe map of all authorized {KSA: GSA} pairs.  This control loop also
+// maintains these pairs in ConfigMap "verifiedSAConfigMapNamespace/verifiedSAConfigMapName".
+type ServiceAccountVerifier struct {
+	hms         *hms.HMSClient
+	saIndexer   cache.Indexer
+	verifiedSAs *saMap
```

Comment:
Can you use the expiring cache? https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/util/cache/expiring.go

Reply:
It works like a cache. Though we use SA event handler to force re-verify (reload).